contracts/utils/TokenCollector.sol

```diff
@@ -32,91 +33,51 @@ abstract contract TokenCollector {
         address from,
         address to,
         uint256 amount,
-        bytes memory data
+        bytes calldata data
     ) internal {
-        (Source src, bytes memory srcData) = abi.decode(data, (Source, bytes));
-        if (src == Source.Token) {
-            return _collectByToken(token, from, to, amount, srcData);
-        }
-        if (src == Source.TokenlonSpender) {
-            return _collectByTokenlonSpender(token, from, to, amount);
-        }
-        if (src == Source.Permit2AllowanceTransfer) {
-            return _collectByPermit2AllowanceTransfer(token, from, to, amount, srcData);
-        }
-        if (src == Source.Permit2SignatureTransfer) {
-            return _collectByPermit2SignatureTransfer(token, from, to, amount, srcData);
-        }
-        revert("TokenCollector: unknown token source");
-    }
+        Source src = Source(uint8(data[0]));
 
-    function _collectByToken(
-        address token,
-        address from,
-        address to,
-        uint256 amount,
-        bytes memory data
-    ) private {
-        if (data.length > 0) {
-            (bool success, bytes memory result) = token.call(abi.encodePacked(IERC20Permit.permit.selector, data));
+        if (src == Source.TokenlonSpender) {
+            ISpender(tokenlonSpender).spendFromUser(from, token, amount);
+            if (to != address(this)) {
+                IERC20(token).safeTransfer(to, amount);
+            }
```

Comment:
This is because we won't be deploying v5 `Spender` to a new version that supports `spendFromUserTo`? I see that there's `spendFromUserTo` in `ISpender` and `Spender`. Maybe we should rollback v5 `Spender` to the older version?

Reply:
Yeah, the spender won't be updated. I think rolling back is a good idea, will be in another PR.